src/input_context.cpp

```diff
@@ -520,84 +555,182 @@ static const std::map<fallback_action, int> fallback_keys = {
     { fallback_action::execute, '.' },
 };
 
+
+keybindings_ui::keybindings_ui( bool permit_execute_action,
+                                input_context *parent ) : cataimgui::window( "KEYBINDINGS", ImGuiWindowFlags_NoNav )
+{
+    filter_text_impl[0] = '\0';
+    this->ctxt = parent;
+
+    legend.push_back( colorize( _( "Unbound keys" ), unbound_key ) );
+    legend.push_back( colorize( _( "Keybinding active only on this screen" ), local_key ) );
+    legend.push_back( colorize( _( "Keybinding active globally" ), global_key ) );
+    if( permit_execute_action ) {
+        legend.push_back( string_format(
+                              _( "Press %c to execute action\n" ),
+                              fallback_keys.at( fallback_action::execute ) ) );
+    }
+    buttons.assign( {
+        {
+            "REMOVE", string_format( _( "[<color_yellow>%c</color>] Remove keybinding" ),
+                                     fallback_keys.at( fallback_action::remove ) )
+        },
+        {
+            "ADD_LOCAL", string_format( _( "[<color_yellow>%c</color>] Add local keybinding" ),
+                                        fallback_keys.at( fallback_action::add_local ) )
+        },
+        {
+            "ADD_GLOBAL", string_format( _( "[<color_yellow>%c</color>] Add global keybinding" ),
+                                         fallback_keys.at( fallback_action::add_global ) )
+        } } );
+}
+
+cataimgui::bounds keybindings_ui::get_bounds()
+{
+    return { -1.f, -1.f, float( str_width_to_pixels( width ) ), float( str_height_to_pixels( TERMY ) ) };
+}
+
+void keybindings_ui::draw_controls()
+{
+    scroll_offset = SIZE_MAX;
+    size_t legend_idx = 0;
+    for( ; legend_idx < 3; legend_idx++ ) {
+        draw_colored_text( legend[legend_idx], c_white );
+        ImGui::SameLine();
+        std::string button_text_no_color = remove_color_tags( buttons[legend_idx].second );
+        ImGui::SetCursorPosX( str_width_to_pixels( width ) - ( get_text_width(
+                                  button_text_no_color ) +
+                              ( ImGui::GetStyle().FramePadding.x * 2 ) + ImGui::GetStyle().ItemSpacing.x ) );
+        action_button( buttons[legend_idx].first, button_text_no_color );
+    }
+    for( ; legend_idx < legend.size(); legend_idx++ ) {
+        draw_colored_text( legend[legend_idx], c_white );
+    }
+    if( last_status != status && status == s_show ) {
+        ImGui::SetKeyboardFocusHere( 0 );
+    }
+    strncpy( filter_text_impl, filter_text.c_str(), strlen( filter_text.c_str() ) );
+    ImGui::InputText( "##NOLABEL", filter_text_impl, 255,
+                      status == s_show ? ImGuiInputTextFlags_None : ImGuiInputTextFlags_ReadOnly );
+    filter_text.assign( filter_text_impl );
```

Comment:
Turns out that 'using std::string with imgui::inputtext' is already a supported use case, unless this fork is older https://stackoverflow.com/a/69046907

Reply:
Also instead of `strlen(filter_text.c_str())` just use `filter_text.length()`.

Reply:
My concern with that example is it appears to pass std::string.c_str() as the buffer to ImGui::InputText, which I *thought* means ImGui will write into the c_str buffer

Reply:
Forgot that you had requested changes here. This should be dealt with, I changed the hard-coded 255 value back to the std::extent<> call I was using before, and I changed the strlen to .length(), which should resolve this